src/main/scala/com/cognite/spark/datasource/TimeSeriesRelation.scala

```diff
@@ -33,301 +30,98 @@ case class TimeSeriesItem(
-      duplicateIds: TimeSeriesConflict): IO[Unit] = {
-    implicit val contextShift: ContextShift[IO] = IO.contextShift(ExecutionContext.global)
+      existingExternalIds: Seq[String],
+      timeSeriesSeq: Seq[TimeSeries]): IO[Unit] = {
+    import CdpConnector.cs
+    val (timeSeriesToUpdate, timeSeriesToCreate) = timeSeriesSeq.partition(
+      p => existingExternalIds.contains(p.externalId.get)
+    )
 
-    val notFound = duplicateIds.notFound
-    val timeSeriesToUpdate = updateTimeSeriesItems.filter(p => !notFound.contains(p.id))
-    val putItems = if (timeSeriesToUpdate.isEmpty) {
-      IO.unit
-    } else {
-      post(
-        config,
-        updateTimeSeriesUrl,
-        timeSeriesToUpdate
-      )
-    }
+    val idMap = client.timeSeries
+      .retrieveByExternalIds(existingExternalIds)
+      .map(_.map(ts => ts.externalId -> ts.id).toMap)
+      .unsafeRunSync()
 
-    val timeSeriesToCreate = timeSeriesItems.filter(p => p.id.exists(notFound.contains(_)))
-    val postTimeSeriesToCreate = timeSeriesToCreate.map(t => PostTimeSeriesItem(t))
-    val postItems = if (timeSeriesToCreate.isEmpty) {
-      IO.unit
-    } else {
-      post(
-        config,
-        timeSeriesUrl,
-        postTimeSeriesToCreate
-      ).flatTap { _ =>
-        IO {
-          if (config.collectMetrics) {
-            timeSeriesCreated.inc(postTimeSeriesToCreate.length)
-          }
-        }
+    val create =
+      if (timeSeriesToCreate.isEmpty) { IO.unit } else {
+        client.timeSeries.createFromRead(timeSeriesToCreate)
+      }
+    val update =
+      if (timeSeriesToUpdate.isEmpty) { IO.unit } else {
+        client.timeSeries.updateFromRead(timeSeriesToUpdate.map(ts =>
+          ts.copy(id = idMap(ts.externalId))))
       }
-    }
 
-    (putItems, postItems).parMapN((_, _) => ())
+    (create, update).parMapN((_, _) => ())
   }
 
+  override def cursors(): Iterator[(Option[String], Option[Int])] =
+    NextCursorIterator[TimeSeriesItem](listUrl("v1"), config, false)
+
   def baseTimeSeriesUrl(project: String, version: String = "v1"): Uri =
     uri"${baseUrl(project, version, config.baseUrl)}/timeseries"
 
-  override def schema: StructType = structType[TimeSeriesItem]
+  override def schema: StructType = structType[TimeSeries]
 
-  override def toRow(t: TimeSeriesItem): Row = asRow(t)
+  override def toRow(t: TimeSeries): Row = asRow(t)
 
-  override def listUrl(): Uri =
-    baseTimeSeriesUrl(config.project, "v1")
-}
+  override def clientToResource(client: GenericClient[IO, Nothing]): TimeSeriesResource[IO] =
+    client.timeSeries
 
+  override def listUrl(version: String): Uri =
+    baseTimeSeriesUrl(config.project, version)
+}
 object TimeSeriesRelation
     extends DeleteSchema
     with UpsertSchema
     with InsertSchema
-    with UpdateSchema {
-  val insertSchema = structType[PostTimeSeriesItem]
-  val upsertSchema = StructType(structType[TimeSeriesItem].filterNot(field =>
+//    with UpdateSchema
```

Comment:
No more update for time series? Coming back in the next PR?

Reply:
Still update for time series! This is just for exposing the schema, which wasn't being used by anyone right now so I skipped writing a special macro for schemas with Setter/NonNullableSetters.

Reply:
Cool :)